plugins/network-elements/nuage-vsp/src/com/cloud/network/manager/NuageVspManagerImpl.java

```diff
@@ -152,50 +243,179 @@ public NuageVspDeviceVO addNuageVspDevice(AddNuageVspDeviceCmd cmd) {
+            paramsTobeUpdated.put("name", "Nuage VSD - " + command.getHostName());
+            paramsTobeUpdated.put("hostname", command.getHostName());
+        }
+
+        if (!Strings.isNullOrEmpty(command.getUserName()) &&
+                !command.getUserName().equals(nuageVspHost.getDetails().get("cmsuser"))) {
+            paramsTobeUpdated.put("cmsuser", command.getUserName());
+        }
+
+        if (!Strings.isNullOrEmpty(command.getPassword())) {
+            String encodedNewPassword = org.apache.commons.codec.binary.StringUtils.newStringUtf8(Base64.encodeBase64(command.getPassword().getBytes()));
+            if (!encodedNewPassword.equals(nuageVspHost.getDetails().get("cmsuserpass"))) {
+                paramsTobeUpdated.put("cmsuserpass", org.apache.commons.codec.binary.StringUtils.newStringUtf8(Base64.encodeBase64(command.getPassword().getBytes())));
+            }
+        }
+
+        if (command.getPort() != null &&
+                command.getPort() != Integer.parseInt(nuageVspHost.getDetails().get("port"))) {
+            paramsTobeUpdated.put("port", String.valueOf(command.getPort()));
+        }
+
+        GetClientDefaultsCommand getClientDefaultsCmd = new GetClientDefaultsCommand();
+        GetClientDefaultsAnswer getClientDefaultsAnswer = (GetClientDefaultsAnswer) _agentMgr.easySend(nuageVspHost.getId(), getClientDefaultsCmd);
+        String apiVersion = MoreObjects.firstNonNull(command.getApiVersion(), getClientDefaultsAnswer.getCurrentApiVersion());
+        SupportedApiVersionCommand supportedApiVersionCmd = new SupportedApiVersionCommand(apiVersion);
+        SupportedApiVersionAnswer supportedApiVersionAnswer = (SupportedApiVersionAnswer) _agentMgr.easySend(nuageVspHost.getId(), supportedApiVersionCmd);
+        if (!supportedApiVersionAnswer.getResult()) {
+            throw new CloudRuntimeException("Incorrect API version: Nuage plugin only supports " + getClientDefaultsAnswer.getCurrentApiVersion());
+        }
+
+        String apiRelativePath = "/nuage/api/" + command.getApiVersion();
+        if (!apiRelativePath.equals(nuageVspHost.getDetails().get("apirelativepath"))) {
+            paramsTobeUpdated.put("apirelativepath", apiRelativePath);
+            paramsTobeUpdated.put("apiversion", apiVersion);
+        }
 
-                        DetailVO detail = new DetailVO(host.getId(), "nuagevspdeviceid", String.valueOf(nuageVspDevice.getId()));
-                        _hostDetailsDao.persist(detail);
+        if (command.getApiRetryCount() != null &&
+                command.getApiRetryCount() != Integer.parseInt(nuageVspHost.getDetails().get("retrycount"))) {
+            paramsTobeUpdated.put("retrycount", String.valueOf(command.getApiRetryCount()));
+        }
+
+        if (command.getApiRetryInterval() != null &&
+                command.getApiRetryInterval() != Integer.parseInt(nuageVspHost.getDetails().get("retryinterval"))) {
```

Comment:
Extract the Integer parsing for a ``final`` local variable to make this ``if`` block easier to read/follow.